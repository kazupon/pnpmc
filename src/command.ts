import { parseArgs, resolveArgs } from 'args-tokens'
import path from 'node:path'
import pc from 'picocolors'
import { commands } from './commands/index.js'
import { fail, log, readPackageJson } from './utils.js'

import type { ArgOptions, ArgToken, ArgValues } from 'args-tokens'
import type { Command, CommandContext, CommandHelpRender } from './commands/types'

const __dirname = path.dirname(new URL(import.meta.url).pathname)

type Commands = keyof typeof commands

const DEFAULT_COMMAND = 'show'

const COMMON_OPTIONS = {
  help: {
    type: 'boolean',
    short: 'h'
  },
  version: {
    type: 'boolean',
    short: 'v'
  }
} as const satisfies ArgOptions

function resolveCommandHelpRender<Options extends ArgOptions>(
  ctx: CommandContext<Options>,
  redner: CommandHelpRender<Options>
): string {
  return typeof redner === 'function' ? redner(ctx) : redner
}

export function showHelp<Options extends ArgOptions>(ctx: CommandContext<Options>): void {
  log(renderHelp(ctx))
}

function renderHelp<Options extends ArgOptions>(ctx: CommandContext<Options>): string {
  const messages = [
    `${resolveCommandHelpRender(ctx, ctx.description)}

USAGE:
${resolveCommandHelpRender(ctx, ctx.help.usage)}

OPTIONS:
${resolveCommandHelpRender(ctx, ctx.help.options)}
`
  ]

  if (ctx.help.examples) {
    messages.push(`EXAMPLES:
${resolveCommandHelpRender(ctx, ctx.help.examples)}`)
  } else {
    messages.push('')
  }

  return messages.join('\n')
}

function getCommandRaw(tokens: ArgToken[]): string {
  const firstToken = tokens[0]
  if (
    firstToken &&
    firstToken.kind === 'positional' &&
    firstToken.index === 0 &&
    firstToken.value
  ) {
    return firstToken.value
  } else {
    return ''
  }
}

async function version(): Promise<string> {
  const pkgJson = await readPackageJson(path.resolve(__dirname, '../package.json'))
  return `${pkgJson.version as string}`
}

async function header(): Promise<string> {
  const pkgJsonPath = path.resolve(__dirname, '../package.json')
  const pkgJson = await readPackageJson(pkgJsonPath)
  const version = pkgJson.version as string
  const name = pkgJson.name as string
  const title = pkgJson.description as string
  return `${title} (${name} v${version})`
}

export function createCommandContext<Options extends ArgOptions, Values = ArgValues<Options>>(
  options: Options,
  values: Values,
  positionals: string[],
  cwd: string,
  command: Command<Options>
): CommandContext<Options, Values> {
  return {
    description: command.description,
    locale: new Intl.Locale('en'), // TODO: resolve locale on runtime and abstraction
    cwd,
    options,
    values,
    positionals,
    help: command.help
  }
}

async function showHelpDefault<Options extends ArgOptions>(
  ctx: CommandContext<Options>
): Promise<void> {
  const loadedCommands = (await Promise.all(
    Object.entries(commands).map(async ([key, cmd]) => [key, await cmd()])
  )) as [Commands, Command<ArgOptions>][]

  const message = `USAGE:
  pnpmc [show] <OPTIONS>
  pnpmc <COMMANDS>

COMMANDS:
${loadedCommands.map(([key, cmd]) => `  ${key.padEnd(21)}${resolveCommandHelpRender(ctx, cmd.description)}`).join('\n')}

For more info, run any command with the \`--help\` flag:
${loadedCommands.map(([key, _]) => `  pnpmc ${key} --help`).join('\n')}

OPTIONS:
${resolveCommandHelpRender(ctx, ctx.help.options)}
  -v, --version        Display the version

EXAMPLES:
${loadedCommands
  .map(([_, cmd]) => cmd.help.examples)
  .filter(Boolean)
  .join('\n\n')}
`
  log(message)
}

function resolveOptions<Options extends ArgOptions>(options: Options): Options {
  return Object.assign(Object.create(null) as Options, options, COMMON_OPTIONS)
}

async function showHeader(): Promise<void> {
  log(pc.cyanBright(await header()))
  log()
}

export async function run(args: string[], cwd: string): Promise<void> {
  const tokens = parseArgs(args)
  const rawCommand = getCommandRaw(tokens)

  const omitted = !rawCommand
  const command = (rawCommand || DEFAULT_COMMAND) as Commands
  const resolvedCommand = (await commands[command]()) as Command<ArgOptions>
  const options = resolveOptions(resolvedCommand.options)

  let resolvedArgs: ReturnType<typeof resolveArgs> | undefined
  try {
    resolvedArgs = resolveArgs(options, tokens)
  } catch (e: unknown) {
    await showHeader()
    if (e instanceof AggregateError) {
      for (const err of e.errors as Error[]) {
        console.error(pc.red(err.message))
      }
      log()
      log(`For more info, run \`pnpmc ${command} --help\` flag`)
      fail()
    } else {
      throw e
    }
  }

  await showHeader()
  const { values, positionals } = resolvedArgs
  if (values.version) {
    log(await version())
    return
  }

  const ctx = createCommandContext(options, values, positionals, cwd, resolvedCommand)
  if (values.help) {
    if (omitted) {
      await showHelpDefault(ctx)
      return
    } else {
      showHelp(ctx)
      return
    }
  }

  await resolvedCommand.run(ctx)
}
