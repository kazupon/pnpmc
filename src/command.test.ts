import { describe, expect, test } from 'vitest'
import { createCommandContext, renderUsage } from './command'

import type { ArgOptions } from 'args-tokens'
import type { Command } from './commands/types'

describe('renderUsage', () => {
  test('basic', () => {
    const command = {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        },
        bar: {
          type: 'boolean'
        },
        baz: {
          type: 'number',
          short: 'b',
          default: 42
        },
        qux: {
          type: 'string',
          short: 'q',
          required: true
        }
      },
      name: 'test',
      description: 'A test command',
      usage: {
        options: {
          foo: 'The foo option',
          bar: 'The bar option',
          baz: 'The baz option',
          qux: 'The qux option'
        },
        examples: `# Example 1\n $test --foo bar --bar --baz 42 --qux quux\n# Example 2\n$ test -f bar -b 42 -q quux`
      },
      run: async () => {
        // something here
      }
    } as Command<ArgOptions>
    const ctx = createCommandContext(
      command.options,
      {}, // argument values
      [],
      { cwd: '/path/to/cmd1', version: '0.0.0', name: 'cmd1' },
      command
    )

    expect(renderUsage(ctx)).toMatchSnapshot()
  })

  test('no options', () => {
    const command = {
      name: 'test',
      description: 'A test command',
      usage: {
        examples: `# Example 1\n $test\n# Example 2\n$ test`
      },
      run: async () => {
        // something here
      }
    } as Command<ArgOptions>
    const ctx = createCommandContext(
      command.options,
      {}, // argument values
      [],
      { cwd: '/path/to/cmd1', version: '0.0.0', name: 'cmd1' },
      command
    )

    expect(renderUsage(ctx)).toMatchSnapshot()
  })

  test('no required options', () => {
    const command = {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        },
        bar: {
          type: 'boolean'
        },
        baz: {
          type: 'number',
          short: 'b',
          default: 42
        }
      },
      name: 'test',
      description: 'A test command',
      usage: {
        options: {
          foo: 'The foo option',
          bar: 'The bar option',
          baz: 'The baz option'
        }
      },
      run: async () => {
        // something here
      }
    } as Command<ArgOptions>
    const ctx = createCommandContext(
      command.options,
      {}, // argument values
      [],
      { cwd: '/path/to/cmd1', version: '0.0.0', name: 'cmd1' },
      command
    )

    expect(renderUsage(ctx)).toMatchSnapshot()
  })

  test('no examples', () => {
    const command = {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        },
        bar: {
          type: 'boolean'
        },
        baz: {
          type: 'number',
          short: 'b',
          default: 42
        },
        qux: {
          type: 'string',
          short: 'q',
          required: true
        }
      },
      name: 'test',
      description: 'A test command',
      usage: {
        options: {
          foo: 'The foo option',
          bar: 'The bar option',
          baz: 'The baz option',
          qux: 'The qux option'
        }
      },
      run: async () => {
        // something here
      }
    } as Command<ArgOptions>
    const ctx = createCommandContext(
      command.options,
      {}, // argument values
      [],
      { cwd: '/path/to/cmd1', version: '0.0.0', name: 'cmd1' },
      command
    )

    expect(renderUsage(ctx)).toMatchSnapshot()
  })

  test('enable usageOptionType', () => {
    const command = {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        },
        bar: {
          type: 'boolean'
        },
        baz: {
          type: 'number',
          short: 'b',
          default: 42
        },
        qux: {
          type: 'string',
          short: 'q',
          required: true
        }
      },
      name: 'test',
      description: 'A test command',
      usage: {
        options: {
          foo: 'The foo option',
          bar: 'The bar option',
          baz: 'The baz option',
          qux: 'The qux option'
        },
        examples: `# Example 1\n $test --foo bar --bar --baz 42 --qux quux\n# Example 2\n$ test -f bar -b 42 -q quux`
      },
      run: async () => {
        // something here
      }
    } as Command<ArgOptions>
    const ctx = createCommandContext(
      command.options,
      {}, // argument values
      [],
      { cwd: '/path/to/cmd1', version: '0.0.0', name: 'cmd1' },
      command,
      { usageOptionType: true, leftMargin: 4, middleMargin: 12 }
    )

    expect(renderUsage(ctx)).toMatchSnapshot()
  })
})

describe.skip('renderUsageDefault', () => {})
