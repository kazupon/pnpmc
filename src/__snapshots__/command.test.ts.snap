// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderUsage > basic 1`] = `
"A test command

USAGE:
  cmd1 test <OPTIONS>

OPTIONS:
  -f, --foo                The foo option
  --bar                    The bar option
  -b, --baz [baz]          The baz option
  -q, --qux <qux>          The qux option

EXAMPLES: 
  # Example 1
   $test --foo bar --bar --baz 42 --qux quux
  # Example 2
  $ test -f bar -b 42 -q quux
"
`;

exports[`renderUsage > enable usageOptionType 1`] = `
"A test command

USAGE:
    cmd1 test <OPTIONS>

OPTIONS:
    -f, --foo                  [string]  The foo option
    --bar                      [boolean] The bar option
    -b, --baz [baz]            [number]  The baz option
    -q, --qux <qux>            [string]  The qux option

EXAMPLES: 
    # Example 1
     $test --foo bar --bar --baz 42 --qux quux
    # Example 2
    $ test -f bar -b 42 -q quux
"
`;

exports[`renderUsage > no examples 1`] = `
"A test command

USAGE:
  cmd1 test <OPTIONS>

OPTIONS:
  -f, --foo                The foo option
  --bar                    The bar option
  -b, --baz [baz]          The baz option
  -q, --qux <qux>          The qux option
"
`;

exports[`renderUsage > no options 1`] = `
"A test command

USAGE:
  cmd1 test 

EXAMPLES: 
  # Example 1
   $test
  # Example 2
  $ test
"
`;

exports[`renderUsage > no required options 1`] = `
"A test command

USAGE:
  cmd1 test [OPTIONS]

OPTIONS:
  -f, --foo                The foo option
  --bar                    The bar option
  -b, --baz [baz]          The baz option
"
`;
